File: src/bot/commands/setup/alias.ts
import {
  MessageFlags,
  PermissionFlagsBits,
  SlashCommandBuilder,
  type ChatInputCommandInteraction
} from 'discord.js';
import { Command } from '../../struct/Command.js';
import { baseEmbed } from '../../../util/embeds.js';
import { EMOJIS } from '../../../util/emojis.js';
import { isValidTag, normalizeTag } from '../../../util/helper.js';

/** Short aliases that stand in for clan tags in commands. */
export default class AliasCommand extends Command {
  public constructor() {
    super('alias', {
      category: 'setup',
      description: 'Manage clan tag aliases.',
      guildOnly: true,
      userPermissions: [PermissionFlagsBits.ManageGuild]
    });
  }

  public builder() {
    return new SlashCommandBuilder()
      .setName('alias')
      .setDescription('Manage clan tag aliases.')
      .setDefaultMemberPermissions(PermissionFlagsBits.ManageGuild)
      .addSubcommand((sub) =>
        sub
          .setName('create')
          .setDescription('Create an alias for a clan tag.')
          .addStringOption((opt) => opt.setName('name').setDescription('Alias name.').setRequired(true))
          .addStringOption((opt) => opt.setName('tag').setDescription('Clan tag.').setRequired(true))
      )
      .addSubcommand((sub) =>
        sub
          .setName('delete')
          .setDescription('Delete an alias.')
          .addStringOption((opt) => opt.setName('name').setDescription('Alias name.').setRequired(true))
      )
      .addSubcommand((sub) => sub.setName('list').setDescription('List all aliases.'));
  }

  public async exec(interaction: ChatInputCommandInteraction, args: Record<string, any>) {
    const guildId = interaction.guildId!;
    switch (args.subcommand) {
      case 'create': {
        if (!isValidTag(args.tag)) {
          return interaction.reply({ content: 'Invalid clan tag.', flags: MessageFlags.Ephemeral });
        }
        const name = String(args.name).toLowerCase();
        await this.client.db.aliases.updateOne(
          { guildId, name },
          { $set: { guildId, name, tag: normalizeTag(args.tag) }, $setOnInsert: { createdAt: new Date() } },
          { upsert: true }
        );
        return interaction.reply(`${EMOJIS.CHECK} Alias \`${name}\` → \`${normalizeTag(args.tag)}\`.`);
      }
      case 'delete': {
        const name = String(args.name).toLowerCase();
        const res = await this.client.db.aliases.deleteOne({ guildId, name });
        return interaction.reply(
          res.deletedCount ? `${EMOJIS.CHECK} Deleted alias \`${name}\`.` : `No alias named \`${name}\`.`
        );
      }
      default: {
        const aliases = await this.client.db.aliases.find({ guildId }).toArray();
        if (!aliases.length) return interaction.reply('No aliases yet. Create one with `/alias create`.');
        const embed = baseEmbed(this.client, interaction)
          .setTitle('Clan Aliases')
          .setDescription(aliases.map((a) => `\`${a.name}\` → \`${a.tag}\``).join('\n'));
        return interaction.reply({ embeds: [embed] });
      }
    }
  }
}

File: cogs/setup/commands/alias.py
"""/alias add|remove|list — short names that stand in for clan tags."""

import discord

from database.db import get_pool
from utils.embeds import base_embed, error_embed, success_embed
from utils.tags import is_valid_tag, normalize_tag


async def alias_add(interaction: discord.Interaction, name: str, tag: str):
    await interaction.response.defer()
    if not is_valid_tag(tag):
        await interaction.followup.send(embed=error_embed("Invalid clan tag."))
        return
    name = name.lower()
    pool = await get_pool()
    await pool.execute(
        """INSERT INTO aliases (guild_id, name, tag) VALUES ($1, $2, $3)
           ON CONFLICT (guild_id, name) DO UPDATE SET tag = EXCLUDED.tag""",
        interaction.guild_id,
        name,
        normalize_tag(tag),
    )
    await interaction.followup.send(embed=success_embed(f"Alias `{name}` → `{normalize_tag(tag)}`."))


async def alias_remove(interaction: discord.Interaction, name: str):
    await interaction.response.defer()
    pool = await get_pool()
    result = await pool.execute("DELETE FROM aliases WHERE guild_id = $1 AND name = $2", interaction.guild_id, name.lower())
    if result.endswith("0"):
        await interaction.followup.send(embed=error_embed(f"No alias named `{name}`."))
        return
    await interaction.followup.send(embed=success_embed(f"Deleted alias `{name}`."))


async def alias_list(interaction: discord.Interaction):
    await interaction.response.defer()
    pool = await get_pool()
    rows = await pool.fetch("SELECT name, tag FROM aliases WHERE guild_id = $1 ORDER BY name", interaction.guild_id)
    if not rows:
        await interaction.followup.send(embed=error_embed("No aliases yet. Create one with `/alias add`."))
        return
    embed = await base_embed(interaction, title="Clan Aliases")
    embed.description = "\n".join(f"`{r['name']}` → `{r['tag']}`" for r in rows)
    await interaction.followup.send(embed=embed)
